### DISCUSSION TITLE
Posibility to select custom midi device

### DISCUSSION THREAD

**Discussion by christiansejer** [shimpe/Integra7AuralAlchemist#3] [Ideas]
Hi
Again. Great project :-)
I have my Integra 7 connected through a Mio XL.
I wonder if you could add the possibility to select the Midi device yourself. For some reason it cannot find my Integra 7 - it is in generic midi mode.
I am on Windows 10 and using the official Roland VST Integra controller you can specify the midi device and this works for my setup.
However, I cannot make your Aural Alchemist find my Integra 7.

Best regards

Christian

**Comment by christiansejer** [comment]
Well. It seems that I cannot even make it connect, when connecting the Integra 7 directly to my PC (not using the Mio XL).
Have tried with Integra in both Generic midi and in Vendor usb mode. No luck...
Has other with Windows 10 made it work for their setup?

Best regards

Christian

**Comment by shimpe** [comment]
Hello there, you have the (questionable?) Honor of being the first living
soul other than me to try it out... in other words, no one has gotten it to
work on anything other than linux. I'm not at the computer now, but i'm
going to see if there's a way i can debug/test it on windows. If you happen
to be a developer as well, any help is welcome 😀.

Op di 19 nov 2024 15:32 schreef christiansejer ***@***.***>:

**Reply by christiansejer** [reply]
My programming skills are rusty, but anything I can do to help?
I guess it somehow has to query for available midi devices, and you can then choose the INTEGRA-7 CTRL device (this works in the official Roland VST plugin).
Do you have any idea how to do that so we can try it out?
Beat regards 
Christian

**Reply by christiansejer** [reply]
I tried to do a little simple debugging by putting out debug text and see at what line i no longer get anything.
It seems that the code stops/fails for me at line 30: "if (!outputs.Any())" in MidiOut.cs. 
So there seems to be issues with getting the output devices properly from the _midiAccessManager as any attempt to use the "var outputs" somehow breaks the code at that point.
Hopefully this can help tracking down the issue.

Best regards

Christian

**Reply by shimpe** [reply]
Hello, thanks for doing some experiments.

Could you please try to do this:

In file MidiOut.cs, between lines 28 and 29 (so just before " var outputs = ..."), add the following code, start the tool and please let me know what the output is:

            foreach (var o  in _midiAccessManager?.Outputs)
            {
                Console.WriteLine(o.Name);
            }

On my system it prints:

Midi Through Port-0
UMC1820 MIDI 1
INTEGRA-7 MIDI 1

But I'm guessing that on yours it will be something else (most likely nothing that includes "INTEGRA-7" in the name). 
If we can confirm that you do see something printed, I will add a midi device selector in one of the coming days (I'm on a tight time schedule these days).

**Reply by christiansejer** [reply]
Thanks for looking into the issue.
Sadly the suggested code does not provide any output and it breaks/stops the code so the following codelines are not executed.
It seems to be an issue with the _midAccessManager under Windows 10. 

Best regards

Christian

**Reply by shimpe** [reply]
Ouch... that is bad news. I'll have to find a windows computer and try it out to see if anything can be remedied.

**Reply by christiansejer** [reply]
Hi again
I got it to work by changing from IMidiAccess2 to IMidiAccess in both MidiOut.cs and MidiIn.cs.
It gives a warning about being obsolete, but it works for me :-)
Now I can do some exploring in the Integra7AuralAlchemist :-)

Best regards

Christian

**Reply by christiansejer** [reply]
Maybe I found the reason?
IMidiAccess2 does not seem to be defined in WinmmMidiAccess.cs https://github.com/atsushieno/managed-midi/blob/master/Commons.Music.Midi.DesktopShared/winmm/WinmmMidiAccess.cs
while it is used in AlsaMidiAccess.cs https://github.com/atsushieno/managed-midi/blob/master/Commons.Music.Midi.DesktopShared/alsaseq/AlsaMidiAccess.cs

Best regards

Christian

**Reply by christiansejer** [reply]
Another hint for Windows users
As the Integra presents both a "INTEGRA-7" and a "INTEGRA-7 CTRL" midi port in Windows it is a good idea to change the line 48: "INTEGRA_CONNECTION_STRING = "INTEGRA-7";" to: "INTEGRA_CONNECTION_STRING = "INTEGRA-7 CTRL";".

This makes it possible to control the Integra while accessing it through the port "INTEGRA-7" in e.g. your DAW. :-)

**Reply by shimpe** [reply]
That is unexpected but great news!  I'll replace IMidiAccess2 with IMidiAccess then, as it also works on Linux that way.

I'm not sure I fully understand the INTEGRA-7 CTRL connection string request. Is the DAW hard-coded to use "INTEGRA-7"? And does INTEGRA-7 always expose two midi devices on windows (so also with the generic MIDI driver e.g.?)
A preferences screen could be added where e.g. the connection string can be reconfigured.

**Reply by christiansejer** [reply]
Great! :-)
On Windows INTEGRA-7 always expose two midi devices. As far as i know Windows only allows one connection to a midi device at a time, so typically one would use the INTEGRA-7 for DAW while the INTEGRA-7 CTRL is used for control.
It is not hard-coded, but common practice.
It would be great with a preference screen with configuration of the connection string. :-)

**Comment by shimpe** [comment]
Hello, I'm again not at the computer (concert season is approaching quickly
😀 and i have some extra rehearsals). Just to rule out the "obvious" first:
is your integra configured to send and receive sysex and midi msgs?

Op wo 20 nov 2024 19:15 schreef christiansejer ***@***.***>:

**Reply by christiansejer** [reply]
Yes. And tested that the Roland vst communication with the Integra works. 
I have it in generic midi, device id 17 and connected through the Mio XL. Did not try with direct connection to the computer this time, as it did not work earlier.
